### PR TITLE
feat(hotspot): new algorithm of hotspot detection

### PR DESCRIPTION
**Pull request by Smityz**
### What problem does this PR solve? 
According to https://github.com/XiaoMi/pegasus/issues/477
### What is changed and how it works?
It is a strategy of hotspot detection https://github.com/XiaoMi/pegasus/pull/444
Compared to the previous algorithm, it makes the following improvements:
1. It defined the concept of threshold, and ready for automatic monitoring.
2. It has higher stability

### Result
1. no hotspot, single set/single get
微信截图_20200221134449
2. one hotspot, single set/single get
微信截图_20200221132757
3. two hotspot, single set/single get
微信截图_20200221144307
4. three hotspot, single set/single get
微信截图_20200221142337
5 one hotspot, multiget

### CODE REVIEW

File: src/server/table_hotspot_policy.h [hunk 1]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
...
+        const auto &anly_data = hotspot_app_data.back();
+        auto temp_data = hotspot_app_data;
+        double avg = 0, sd = 0;
+        int n = 0;
+        // avg: Average number
+        // sd: Standard deviation
+        // n: Number of samples
+        while (!temp_data.empty()) {
+            for (auto partition_data : temp_data.front()) {
+                if (partition_data.total_qps > 1.00) {
```
Comment (levy5307) [inline src/server/table_hotspot_policy.h]: double类型不能直接比较大小

File: src/server/table_hotspot_policy.h [hunk 2]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
...
+class hotspot_algo_qps_variance : public hotspot_policy
+{
+public:
+    void analysis(const std::queue<std::vector<hotspot_partition_data>> &hotspot_app_data,
+                  std::vector<::dsn::perf_counter_wrapper> &hot_points)
+    {
+        dassert(hotspot_app_data.back().size() == hot_points.size(),
+                "partition counts error, please check");
+        std::vector<double> data_samples;
+        data_samples.reserve(hotspot_app_data.size() * hot_points.size());
```
Comment (levy5307) [inline src/server/table_hotspot_policy.h]:
```
std::vector<double> data_samples(hotspot_app_data.size() * hot_points.size());
```

File: src/server/table_hotspot_policy.h [hunk 3]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
...
+        double avg = 0, sd = 0;
+        int n = 0;
+        // avg: Average number
+        // sd: Standard deviation
+        // n: Number of samples
+        while (!temp_data.empty()) {
+            for (auto partition_data : temp_data.front()) {
+                if (partition_data.total_qps - 1.00 > 0) {
+                    data_samples.push_back(partition_data.total_qps);
+                    avg += partition_data.total_qps;
```
Comment (acelyc111) [inline src/server/table_hotspot_policy.h]: `avg` is wrong named here, you can define a new variable when you need it.

File: src/server/table_hotspot_policy.h [hunk 4]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
     }
 };
 
+// PauTa Criterion
+class hotspot_algo_qps_variance : public hotspot_policy
+{
+public:
+    void analysis(const std::queue<std::vector<hotspot_partition_data>> &hotspot_app_data,
+                  std::vector<::dsn::perf_counter_wrapper> &hot_points)
```
Comment (acelyc111) [inline src/server/table_hotspot_policy.h]: Rename `hot_points ` to `perf_counters` would be better.

File: src/server/table_hotspot_policy.h [hunk 5]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
...
+    }
+};
+
 // hotspot_calculator is used to find the hotspot in Pegasus
 class hotspot_calculator
 {
 public:
     hotspot_calculator(const std::string &app_name, const int partition_num)
-        : _app_name(app_name), _points(partition_num), _policy(new hotspot_algo_qps_skew())
+        : _app_name(app_name), _points(partition_num), _policy(new hotspot_algo_qps_variance())
```
Comment (acelyc111) [inline src/server/table_hotspot_policy.h]: 1. Read `hotspot_policy` name from config file
2. Create a `hotspot_policy` object according to the name
Then you can change hotspot_policy algorithm by update config, no need to recompile.

File: src/server/test/pegasus_tablehotspot_test.cpp [hunk 6]
```diff
@@ -9,19 +9,25 @@
 namespace pegasus {
 namespace server {
 
-TEST(table_hotspot_policy, hotspot_algo_qps_skew)
+TEST(table_hotspot_policy, hotspot_algo_qps_variance)
```
Comment (acelyc111) [inline src/server/test/pegasus_tablehotspot_test.cpp]: The previous algorithm is missing test now

File: src/server/table_hotspot_policy.h [hunk 7]
```diff
@@ -42,12 +44,59 @@ class hotspot_algo_qps_skew : public hotspot_policy
...
+        const auto &anly_data = hotspot_app_data.back();
+        auto temp_data = hotspot_app_data;
+        double avg = 0, sd = 0;
+        int n = 0;
+        // avg: Average number
+        // sd: Standard deviation
+        // n: Number of samples
+        while (!temp_data.empty()) {
+            for (auto partition_data : temp_data.front()) {
+                if (partition_data.total_qps - 1.00 > 0) {
```
[the remark this answers is not in the document]
Comment (Smityz) [reply src/server/table_hotspot_policy.h]: The value less than 1.00, is invalid data and will affect statistical results

File: src/server/info_collector.cpp [hunk 8]
```diff
@@ -285,7 +289,14 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
     if (iter != _hotspot_calculator_store.end()) {
         return iter->second;
     }
-    hotspot_calculator *calculator_address = new hotspot_calculator(app_name, partition_num);
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_detect_algorithm") {
```
Comment (acelyc111) [inline src/server/info_collector.cpp]: Should compare with `hotspot_algo_qps_variance `?

File: src/server/info_collector.cpp [hunk 9]
```diff
@@ -285,7 +289,17 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
     }
-    hotspot_calculator *calculator_address = new hotspot_calculator(app_name, partition_num);
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_algo_qps_variance") {
+        policy.reset(new hotspot_algo_qps_variance());
+    } else if (_hotspot_detect_algorithm == "hotspot_algo_qps_skew") {
+        policy.reset(new hotspot_algo_qps_skew());
+    } else {
+        ddebug("no such hotspot detect algorithm");
+        policy.reset(new hotspot_algo_qps_variance());
```
Comment (acelyc111) [inline src/server/info_collector.cpp]: 这里就disable掉吧

**Comment by acelyc111** [comment]
CI失败需要看下

File: src/server/test/config.ini [hunk 10]
```diff
@@ -503,5 +503,5 @@ onebox = 1
 onebox2 = 2
 
 [pegasus.clusters]
-onebox = @LOCAL_IP@:34701,@LOCAL_IP@:34702,@LOCAL_IP@:34703
-onebox2 = @LOCAL_IP@:34701,@LOCAL_IP@:34702,@LOCAL_IP@:34703
+onebox = 0.0.0.0:34701,0.0.0.0:34702,0.0.0.0:34703
+onebox2 = 0.0.0.0:34701,0.0.0.0:34702,0.0.0.0:34703
```
Comment (acelyc111) [inline src/server/test/config.ini]: Remove these code, the problem has been solved by your another PR.

File: src/server/info_collector.cpp [hunk 11]
```diff
@@ -285,7 +289,17 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
     }
-    hotspot_calculator *calculator_address = new hotspot_calculator(app_name, partition_num);
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_algo_qps_variance") {
+        policy.reset(new hotspot_algo_qps_variance());
+    } else if (_hotspot_detect_algorithm == "hotspot_algo_qps_skew") {
+        policy.reset(new hotspot_algo_qps_skew());
+    } else {
+        ddebug("no such hotspot detect algorithm");
+        policy.release();
```
Comment (acelyc111) [inline src/server/info_collector.cpp]: `policy` is not initialized, not needed to release.

File: src/server/info_collector.cpp [hunk 12]
```diff
@@ -285,7 +289,17 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
+    }
+    hotspot_calculator *calculator_address =
```
Comment (acelyc111) [inline src/server/info_collector.cpp]: `calculator_address ` -> `calculator`

File: src/server/info_collector.cpp (hunk 11 above)
Comment (acelyc111) [inline src/server/info_collector.cpp]: Return `nullptr` here, and you should check whether it work when `get_hotspot_calculator` return `nullptr`, may add some protective code.

File: src/server/info_collector.cpp [hunk 13]
```diff
@@ -285,9 +293,20 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
-    return calculator_address;
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_algo_qps_variance") {
+        policy.reset(new hotspot_algo_qps_variance());
+    } else if (_hotspot_detect_algorithm == "hotspot_algo_qps_skew") {
+        policy.reset(new hotspot_algo_qps_skew());
+    } else {
+        ddebug("no such hotspot detect algorithm");
+        policy = nullptr;
+        return NULL;
```
Comment (acelyc111) [inline src/server/info_collector.cpp]:
```suggestion
        return nullptr;
```

File: src/server/info_collector.cpp [hunk 14]
```diff
@@ -285,9 +293,20 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
-    hotspot_calculator *calculator_address = new hotspot_calculator(app_name, partition_num);
-    _hotspot_calculator_store[app_name] = calculator_address;
-    return calculator_address;
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_algo_qps_variance") {
+        policy.reset(new hotspot_algo_qps_variance());
+    } else if (_hotspot_detect_algorithm == "hotspot_algo_qps_skew") {
+        policy.reset(new hotspot_algo_qps_skew());
+    } else {
+        ddebug("no such hotspot detect algorithm");
```
Comment (acelyc111) [inline src/server/info_collector.cpp]:
```suggestion
        dwarn("hotspot detection is disabled");
```

File: src/server/info_collector.cpp [hunk 15]
```diff
@@ -148,6 +152,10 @@ void info_collector::on_app_stat()
         // hotspot_calculator is to detect hotspots
         hotspot_calculator *hotspot_calculator =
             get_hotspot_calculator(app_rows.first, app_rows.second.size());
+        if (hotspot_calculator == NULL) {
+            ddebug("hotspot_calculator pointer is NULL, hotspot detect won't continue!");
```
Comment (acelyc111) [inline src/server/info_collector.cpp]:
```suggestion
        if (!hotspot_calculator) {
```

File: src/server/info_collector.cpp [hunk 16]
```diff
@@ -285,9 +292,19 @@ hotspot_calculator *info_collector::get_hotspot_calculator(const std::string &ap
...
-    hotspot_calculator *calculator_address = new hotspot_calculator(app_name, partition_num);
-    _hotspot_calculator_store[app_name] = calculator_address;
-    return calculator_address;
+    std::unique_ptr<hotspot_policy> policy;
+    if (_hotspot_detect_algorithm == "hotspot_algo_qps_variance") {
+        policy.reset(new hotspot_algo_qps_variance());
+    } else if (_hotspot_detect_algorithm == "hotspot_algo_qps_skew") {
+        policy.reset(new hotspot_algo_qps_skew());
+    } else {
+        dwarn("hotspot detection is disabled");
```
Comment (acelyc111) [inline src/server/info_collector.cpp]:
```suggestion
        dwarn("hotspot detection is disabled");
        _hotspot_calculator_store[app_name] = nullptr;
```

File: src/server/table_hotspot_policy.h [hunk 17]
```diff
@@ -20,24 +22,71 @@ class hotspot_policy
...
+    void analysis(const std::queue<std::vector<hotspot_partition_data>> &hotspot_app_data,
+                  std::vector<::dsn::perf_counter_wrapper> &perf_counters)
+    {
+        dassert(hotspot_app_data.back().size() == perf_counters.size(),
+                "partition counts error, please check");
+        std::vector<double> data_samples;
+        data_samples.reserve(hotspot_app_data.size() * perf_counters.size());
+        auto temp_data = hotspot_app_data;
+        double total = 0, sd = 0, avg = 0;
+        int n = 0;
```
Comment (levy5307) [inline src/server/table_hotspot_policy.h]: 这个n建议改成一个比较直观的名字，例如：sample_count

**Pull request merged**